kitsune/wiki/facets.py

```diff
@@ -73,62 +77,43 @@ def documents_for(locale, topics=None, products=None):
-        # but they won't be in the correct sort (by votes in the last
-        # 30 days). It is better to return them in the wrong order
-        # than not to return them at all.
-        documents = _db_documents_for(locale, topics, products)
-        cache.add(_documents_for_cache_key(locale, topics, products), documents)
-
-    return documents
-
-
-def _es_documents_for(locale, topics=None, products=None):
-    """ES implementation of documents_for."""
-    s = (
-        DocumentMappingType.search()
-        .values_dict("id", "document_title", "url", "document_parent_id", "document_summary")
-        .filter(
-            document_locale=locale,
-            document_is_archived=False,
-            document_category__in=settings.IA_DEFAULT_CATEGORIES,
-        )
+    qs = (
+        Document.objects.filter(
+            locale=locale,
+            is_archived=False,
+            current_revision__isnull=False,
+            category__in=settings.IA_DEFAULT_CATEGORIES,
+        ).select_related("current_revision", "parent")
+        # speed up query by removing any ordering, since we're doing it in python:
+        .order_by()
     )
+    if topics:
+        qs = qs.filter(Q(topics__in=topics) | Q(parent__topics__in=topics))
+    if products:
+        qs = qs.filter(Q(products__in=products) | Q(parent__products__in=products))
+
+    votes_cache_key = f"votes_for:{cache_key}"
+    votes_dict = cache.get(votes_cache_key)
+    if votes_dict is None:
+        votes_query = (
+            HelpfulVote.objects.filter(
+                revision_id__in=qs.values_list("current_revision_id", flat=True),
+                created__gt=datetime.now() - timedelta(days=30),
+                helpful=True,
+            )
+            .values("revision_id")
+            .annotate(count=Count("*"))
+            .values("revision_id", "count")
+        )
+        votes_dict = {row["revision_id"]: row["count"] for row in votes_query}
+        cache.set(votes_cache_key, votes_dict, timeout=VOTES_CACHE_TIMEOUT)
 
-    for topic in topics or []:
-        s = s.filter(topic=topic.slug)
-    for product in products or []:
-        s = s.filter(product=product.slug)
-
-    results = s.order_by("document_display_order", "-document_recent_helpful_votes")[:100]
-    results = DocumentMappingType.reshape(results)
-    return results
-
-
-def _db_documents_for(locale, topics=None, products=None):
-    """DB implementation of documents_for."""
-    qs = Document.objects.filter(
-        locale=locale,
-        is_archived=False,
-        current_revision__isnull=False,
-        category__in=settings.IA_DEFAULT_CATEGORIES,
-    ).prefetch_related("current_revision")
-    for topic in topics or []:
-        qs = qs.filter(topics=topic)
-    for product in products or []:
-        qs = qs.filter(products=product)
-
-    # Convert the results to a dicts to look like the ES results.
     doc_dicts = []
     for d in qs.distinct():
```

Comment:
Do we need this here? Removing it will benefit performance. I also wonder if we can use `set` in case we do need it. Testing this locally, `set` was faster by 5ms on average

Reply:
Thinking on this, the query should be always of a certain size and not return lots of results. Thoughts?

Reply:
Yeah I don't think we need the distinct how we're now using this query